{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Build Pedigree Graph from Nodelist and Edgelist"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "import networkx as nx\n",
    "\n",
    "G = nx.DiGraph()\n",
    "\n",
    "with open(\"nodelist.txt\") as f:\n",
    "    nodes = f.read().splitlines()\n",
    "\n",
    "with open(\"edgelist.txt\") as f:\n",
    "    edges = f.read().splitlines()\n",
    "\n",
    "for node in nodes: \n",
    "\n",
    "    if node == \"filename,node,gen,meth\" or node == \"\":\n",
    "        continue\n",
    "    \n",
    "    filename, node, gen, meth = node.split(',')\n",
    "    \n",
    "    node = {\n",
    "        'filename': filename,\n",
    "        'node': node,\n",
    "        'gen': int(gen),\n",
    "        'meth': True if meth == 'Y' else False\n",
    "    }\n",
    "\n",
    "    \n",
    "\n",
    "    G.add_node(node['node'], **node)\n",
    "\n",
    "for edge in edges:\n",
    "\n",
    "    if edge == \"from,to\":\n",
    "        continue\n",
    "\n",
    "    from_, to_ = edge.split(',')\n",
    "\n",
    "    G.add_edge(from_, to_)\n",
    "\n",
    "def get_predecessor_node(node):\n",
    "    if not node in G:\n",
    "        return None\n",
    "\n",
    "    pred = iter(G.pred[node])\n",
    "\n",
    "    if pred.__length_hint__() == 0:\n",
    "        return None\n",
    "\n",
    "    pred = G.nodes[next(pred)]\n",
    "\n",
    "    if pred[\"meth\"]:\n",
    "        return pred\n",
    "    else:\n",
    "        return get_predecessor_node(pred['node'])\n",
    "\n",
    "def get_pred_node_by_gen_and_line(gen, line):\n",
    "    pred =  get_predecessor_node(f\"{gen}_{line}\")\n",
    "    if pred is None:\n",
    "        return None, None\n",
    "    gen, line = pred[\"node\"].split('_')\n",
    "    return int(gen), int(line)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Hyperparamters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "import torch\n",
    "\n",
    "batch_size = 32\n",
    "\n",
    "cs_neighbours = 5\n",
    "hist_mod_neighbours = 20\n",
    "gene_neighbours = 1\n",
    "meth_neighbours = 20\n",
    "\n",
    "mutables = 5 # not implemented yet\n",
    "\n",
    "device = \"cuda\" if torch.cuda.is_available() else \"cpu\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "import torch\n",
    "import clickhouse_connect\n",
    "import polars as pl\n",
    "import time\n",
    "\n",
    "class Sites(torch.utils.data.IterableDataset):\n",
    "    def __init__(self, mode=\"train\"):\n",
    "        super(Sites).__init__()\n",
    "        self.mode = mode\n",
    "\n",
    "        self.client = clickhouse_connect.get_client(host='localhost', username='cgoeldel', password='Goe1409ldel')\n",
    "        genes = self.client.query_arrow(f\"select * from genes where chromosome < 5\") if mode == \"train\" else self.client.query_arrow(f\"select * from genes where chromosome = 5\")\n",
    "        self.genes = pl.DataFrame(genes)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "        # self.samples_per_chrom = pl.DataFrame(self.client.query_arrow(\"select chromosome, count(*) from methylome_within_gbM where generation = 0 group by chromosome\").result_rows[0][0])\n",
    "        # self.samples = self.client.query(\"select count(distinct(generation, line)) from methylome_within_gbM\").results_rows[0][0]\n",
    "        # self.hist_mod_dict = {  \n",
    "        #         \"input\" : 1,\n",
    "        #         \"H3\" : 2,\n",
    "        #         \"H3K4Me1\" : 3,\n",
    "        #         \"H3K27Me3\" : 4,\n",
    "        #         \"H2AZ\" : 5,\n",
    "        #         \"H3K56Ac\" : 6,\n",
    "        #         \"H3K4Me3\" : 7    \n",
    "        #         }\n",
    "\n",
    "    \n",
    "    def __iter__(self):\n",
    "        for gene in self.genes.iter_rows(named=True):\n",
    "            tic = time.perf_counter()\n",
    "\n",
    "            # TODO: more elegant way to iterate over all pedigree branches\n",
    "            gene_length = gene['end'] - gene['start']\n",
    "            sites_line_2 = self.client.query_arrow(f\"select location - {gene['start']} as start_diff, {gene['end']} -location as end_diff, location * 1.0 / ({gene_length}) as percentile, meth_lvl from methylome_within_gbM where chromosome = {gene['chromosome']} and (line = 2 or line = 0)  and strand = {gene['strand']} and location between {gene['start']} and {gene['end']} order by generation, line, start_diff\")\n",
    "            sites_line_2 = pl.DataFrame(sites_line_2)\n",
    "            # sites_line_8 = self.client.query_arrow(f\"select location - {gene['start']} as start_diff, {gene['end']} -location as end_diff, location * 1.0 / ({gene_length}) as percentile, meth_lvl from methylome_within_gbM where chromosome = {gene['chromosome']} and (line = 8 or line = 0)  and strand = {gene['strand']} and location between {gene['start']} and {gene['end']} order by generation, line, start_diff\")\n",
    "            # sites_line_8 = pl.DataFrame(sites_line_8)\n",
    "            print(sites_line_2.shape)\n",
    "            display(gene)\n",
    "       \n",
    "            targets_line_2 = sites_line_2['meth_lvl'].to_numpy()\n",
    "            # targets_line_8 = sites_line_8['meth_lvl'].to_numpy()\n",
    "\n",
    "            targets_line_2 = torch.tensor(targets_line_2, dtype=torch.float32)\n",
    "            # targets_line_8 = torch.tensor(targets_line_8, dtype=torch.float32)\n",
    "\n",
    "            targets_line_2 = targets_line_2.reshape(gene_length, 6)\n",
    "            # targets_line_8 = targets_line_8.reshape(gene_length, 6)\n",
    "\n",
    "            targets_line_2 = targets_line_2[1:, :] # remove first row, these can not be predicted as there is no predecessor\n",
    "            # targets_line_8 = targets_line_8[1:, :]\n",
    "            \n",
    "            sites_line_2 = torch.tensor(sites_line_2.to_numpy(), dtype=torch.float32).reshape(gene_length * 4, 6)\n",
    "            # sites_line_8 = torch.tensor(sites_line_8.to_numpy(), dtype=torch.float32).reshape(gene_length, 6)\n",
    "\n",
    "            # remove the last row, these have no predecessor, so no known target\n",
    "            sites_line_2 = sites_line_2[:-1, :]\n",
    "            # sites_line_8 = sites_line_8[:-1, :]\n",
    "\n",
    "            chr_states = self.client.query_arrow(f\"select state, ({gene['start']} - start ) as start_diff, ({gene['end']} - end) as end_diff from chr_states where chromosome = {gene['chromosome']} and start between {gene['start']} and {gene['end']} and end between {gene['start']} and {gene['end']}\")\n",
    "            chr_states = pl.DataFrame(chr_states)\n",
    "\n",
    "            chr_states_and_gene = chr_states.vstack(gene)\n",
    "\n",
    "            chr_states = torch.tensor(chr_states_and_gene.to_numpy(), dtype=torch.float32)\n",
    "\n",
    "            print(sites_line_2.shape)\n",
    "            # print(sites_line_8.shape)\n",
    "            print(chr_states.shape)\n",
    "            print(targets_line_2.shape)\n",
    "            # print(targets_line_8.shape)\n",
    "\n",
    "            # every last node -> list back to root: decoder \n",
    "            # surrounding methylome -> encoder\n",
    "            # surrounding gene + chromatine state -> encoder\n",
    "            x = sites_line_2.to(device)\n",
    "            c = chr_states.to(device)\n",
    "            t = targets_line_2.to(device)\n",
    "\n",
    "\n",
    "            toc = time.perf_counter()\n",
    "            print(f\"Gene {gene['chromosome']}:{gene['start']}:{gene['end']} in {toc - tic:0.4f} seconds\")\n",
    "            yield x, c, t\n",
    "        \n",
    "   \n",
    "   \n",
    "   \n",
    "   \n",
    "   \n",
    "    \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def save_surroundings(chromosome, location, idx, targets, mode=\"train\"): \n",
    "    idx = idx.tolist()\n",
    "    targets = targets.tolist()\n",
    "    data = [chromosome, location, idx, targets, mode]\n",
    "    \n",
    "    client.insert(\"training_data\", [data], column_names=[\"chromosome\", \"location\", \"prompt\", \"targets\", \"mode\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_neighbours(site, mode=\"train\"):\n",
    "    # get surrounding genes\n",
    "    tic = time.perf_counter()\n",
    "    genes = client.query_arrow(f\"select type, ({site['location']} - start) as start_diff, ({site['location']} - end) as end_diff from genes where chromosome = {site['chromosome']} and strand = {site['strand']} order by abs(start_diff) limit {gene_neighbours}\")\n",
    "    g = pl.DataFrame(genes)\n",
    "\n",
    "    toc = time.perf_counter()\n",
    "    # print(f\"genes in {toc - tic:0.4f} seconds\")\n",
    "    # display(g)\n",
    "    # Get surrounding histone mods \n",
    "    tic = time.perf_counter()\n",
    "\n",
    "    offset = 200\n",
    "    sufficient = False\n",
    "    while not sufficient:\n",
    "        histone_mods = client.query_arrow(f\"select modification, ({site['location']} - start ) as start_diff, ({site['location']} - end) as end_diff from histone_mods where chromosome = {site['chromosome']} and start > {site['location'] - offset} and end < {site['location'] + offset} order by abs(start_diff) limit {hist_mod_neighbours}\")\n",
    "        h = pl.DataFrame(histone_mods)\n",
    "        if len(h) == hist_mod_neighbours:\n",
    "            sufficient = True\n",
    "        else:\n",
    "            print(f\"Only {len(h)} histone mods found, increasing offset to {offset * 2}\")\n",
    "            offset *= 2\n",
    "            continue\n",
    "\n",
    "        h = h.with_columns(pl.col(\"modification\").map_dict(hist_mod_dict).alias(\"modification\"))\n",
    "    \n",
    "    toc = time.perf_counter()\n",
    "    # print(f\"Histone mods in {toc - tic:0.4f} seconds\")\n",
    "    # display(h)\n",
    "    # Get surrounding chromatine states\n",
    "    tic = time.perf_counter()\n",
    "    chr_states = client.query_arrow(f\"select state, ({site['location']} - start ) as start_diff, ({site['location']} -end) as end_diff from chr_states where chromosome = {site['chromosome']} order by abs(start_diff) limit {cs_neighbours } \")\n",
    "    c = pl.DataFrame(chr_states)\n",
    "\n",
    "    toc = time.perf_counter()\n",
    "    # print(f\"Chromatine State in {toc - tic:0.4f} seconds\")\n",
    "    # display(c)\n",
    "    # Get each site in all generations and lines \n",
    "    tic = time.perf_counter()\n",
    "    offset = 200\n",
    "    sufficient = False\n",
    "    while not sufficient:\n",
    "        all_generations_and_neighbours = client.query_arrow(f\"select * except (trinucleotide_context, pedigree, id), ({site['location']} - location ) as location_diff from methylome where chromosome = {site['chromosome']} and strand = {site['strand']} and location between  {site['location'] - offset} and  {site['location'] + offset} order by abs(location_diff), location_diff, generation, line limit {(meth_neighbours +1)  * samples}\")\n",
    "        m = pl.DataFrame(all_generations_and_neighbours) # (meth_neighbours * samples, 12)\n",
    "        if len(m) == (meth_neighbours +1)  * samples:\n",
    "            sufficient = True\n",
    "        else:\n",
    "            print(f\"Only {len(m)} neighbours found, increasing offset to {offset * 2}\")\n",
    "            offset *= 2\n",
    "            continue\n",
    "    # display(m)\n",
    "\n",
    "    toc = time.perf_counter()\n",
    "    # print(f\"neighbours in {toc - tic:0.4f} seconds\")                \n",
    "    tic = time.perf_counter()\n",
    "    site_across_generations = m.filter(pl.col(\"location_diff\") == 0)\n",
    "    m = m.filter(pl.col(\"location_diff\") != 0)\n",
    "   \n",
    "    m = m.filter((pl.col(\"generation\") != 0) & (pl.col(\"line\") != 0))\n",
    "\n",
    "    preds = []\n",
    "    targets = []\n",
    "    for site in site_across_generations.iter_rows(named=True):\n",
    "        pred_gen, pred_line = get_pred_node_by_gen_and_line(site[\"generation\"], site[\"line\"])\n",
    "        if not pred_gen is None:\n",
    "            pred = site_across_generations.filter((pl.col(\"generation\") == pred_gen) & (pl.col(\"line\") == pred_line)) # (1, 12)\n",
    "            preds.append(torch.tensor(pred.to_numpy()[0], dtype=torch.float32))\n",
    "            targets.append(site[\"meth_lvl\"])\n",
    "\n",
    "    toc = time.perf_counter()\n",
    "    # print(f\"Filtering for {toc - tic:0.4f} seconds\")    \n",
    "    tic = time.perf_counter()\n",
    "    t = torch.tensor(targets, dtype=torch.float32)\n",
    "    p = torch.stack(preds) # (samples - 1, 12)\n",
    "    g = torch.tensor(g.to_numpy(), dtype=torch.float32) \n",
    "    m = torch.tensor(m.to_numpy(),dtype=torch.float32).T \n",
    "    h = torch.tensor(h.to_numpy(),dtype=torch.float32)\n",
    "    c = torch.tensor(c.to_numpy(),dtype=torch.float32)\n",
    "\n",
    "    m = m.reshape((samples-1), 12 * meth_neighbours) # (samples -1, concatenated neighbours)\n",
    "    \n",
    "    # These are time-invariant\n",
    "    g = g.reshape(3 * gene_neighbours)\n",
    "    h = h.reshape(3 * hist_mod_neighbours)\n",
    "    c = c.reshape(3* cs_neighbours)\n",
    "\n",
    "    # surroundings \n",
    "    s = torch.cat([g, c, h])\n",
    "    s = s.expand((samples -1), -1)\n",
    "\n",
    "    x = torch.cat([m, p, s], dim=1)\n",
    "\n",
    "    toc = time.perf_counter()\n",
    "    # print(f\"Reshaping in {toc - tic:0.4f} seconds\")\n",
    "    save_surroundings(site[\"chromosome\"], site[\"location\"], x, t, mode)\n",
    "\n",
    "    return x, t\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Creating 254372 samples in range 0 to 1000000.0\n",
      "Creating sample 0 in mode test at location 109 on chromosome 1\n",
      "Created sample 0 in 0.0418 seconds\n",
      "Creating sample 1 in mode train at location 110 on chromosome 1\n",
      "Created sample 1 in 0.0423 seconds\n",
      "Creating sample 2 in mode train at location 115 on chromosome 1\n",
      "Created sample 2 in 0.0364 seconds\n",
      "Creating sample 3 in mode train at location 116 on chromosome 1\n",
      "Created sample 3 in 0.0385 seconds\n",
      "Creating sample 4 in mode train at location 161 on chromosome 1\n",
      "Created sample 4 in 0.0372 seconds\n",
      "Creating sample 5 in mode train at location 162 on chromosome 1\n",
      "Created sample 5 in 0.0389 seconds\n",
      "Creating sample 6 in mode train at location 310 on chromosome 1\n",
      "Created sample 6 in 0.0380 seconds\n",
      "Creating sample 7 in mode train at location 311 on chromosome 1\n",
      "Created sample 7 in 0.0389 seconds\n",
      "Creating sample 8 in mode train at location 500 on chromosome 1\n",
      "Created sample 8 in 0.0375 seconds\n",
      "Creating sample 9 in mode train at location 501 on chromosome 1\n",
      "Created sample 9 in 0.0384 seconds\n",
      "Creating sample 10 in mode train at location 511 on chromosome 1\n",
      "Created sample 10 in 0.0356 seconds\n",
      "Creating sample 11 in mode train at location 512 on chromosome 1\n",
      "Created sample 11 in 0.0405 seconds\n",
      "Creating sample 12 in mode train at location 642 on chromosome 1\n",
      "Created sample 12 in 0.0366 seconds\n",
      "Creating sample 13 in mode train at location 643 on chromosome 1\n",
      "Created sample 13 in 0.0379 seconds\n",
      "Creating sample 14 in mode train at location 647 on chromosome 1\n",
      "Created sample 14 in 0.0351 seconds\n",
      "Creating sample 15 in mode train at location 648 on chromosome 1\n",
      "Created sample 15 in 0.0379 seconds\n",
      "Creating sample 16 in mode validation at location 650 on chromosome 1\n",
      "Created sample 16 in 0.0364 seconds\n",
      "Creating sample 17 in mode test at location 651 on chromosome 1\n",
      "Created sample 17 in 0.0392 seconds\n",
      "Creating sample 18 in mode train at location 790 on chromosome 1\n",
      "Created sample 18 in 0.0355 seconds\n",
      "Creating sample 19 in mode train at location 791 on chromosome 1\n",
      "Created sample 19 in 0.0391 seconds\n",
      "Creating sample 20 in mode train at location 809 on chromosome 1\n",
      "Created sample 20 in 0.0362 seconds\n",
      "Creating sample 21 in mode train at location 810 on chromosome 1\n",
      "Created sample 21 in 0.0379 seconds\n",
      "Creating sample 22 in mode train at location 839 on chromosome 1\n",
      "Created sample 22 in 0.0368 seconds\n",
      "Creating sample 23 in mode train at location 840 on chromosome 1\n",
      "Created sample 23 in 0.0385 seconds\n",
      "Creating sample 24 in mode train at location 904 on chromosome 1\n",
      "Created sample 24 in 0.0384 seconds\n",
      "Creating sample 25 in mode validation at location 905 on chromosome 1\n",
      "Created sample 25 in 0.0383 seconds\n",
      "Creating sample 26 in mode train at location 911 on chromosome 1\n",
      "Created sample 26 in 0.0366 seconds\n",
      "Creating sample 27 in mode train at location 912 on chromosome 1\n",
      "Created sample 27 in 0.0395 seconds\n",
      "Creating sample 28 in mode train at location 948 on chromosome 1\n",
      "Created sample 28 in 0.0364 seconds\n",
      "Creating sample 29 in mode train at location 949 on chromosome 1\n",
      "Created sample 29 in 0.0448 seconds\n",
      "Creating sample 30 in mode train at location 1324 on chromosome 1\n",
      "Created sample 30 in 0.0374 seconds\n",
      "Creating sample 31 in mode train at location 1325 on chromosome 1\n",
      "Created sample 31 in 0.0430 seconds\n",
      "Creating sample 32 in mode test at location 1439 on chromosome 1\n",
      "Created sample 32 in 0.0421 seconds\n",
      "Creating sample 33 in mode test at location 1440 on chromosome 1\n",
      "Created sample 33 in 0.0522 seconds\n",
      "Creating sample 34 in mode train at location 1541 on chromosome 1\n",
      "Created sample 34 in 0.0477 seconds\n",
      "Creating sample 35 in mode train at location 1542 on chromosome 1\n",
      "Created sample 35 in 0.0385 seconds\n",
      "Creating sample 36 in mode train at location 1634 on chromosome 1\n",
      "Created sample 36 in 0.0405 seconds\n",
      "Creating sample 37 in mode train at location 1635 on chromosome 1\n",
      "Created sample 37 in 0.0377 seconds\n",
      "Creating sample 38 in mode train at location 1649 on chromosome 1\n",
      "Created sample 38 in 0.0371 seconds\n",
      "Creating sample 39 in mode train at location 1650 on chromosome 1\n",
      "Created sample 39 in 0.0383 seconds\n",
      "Creating sample 40 in mode train at location 1778 on chromosome 1\n",
      "Created sample 40 in 0.0461 seconds\n",
      "Creating sample 41 in mode train at location 1779 on chromosome 1\n",
      "Created sample 41 in 0.0412 seconds\n",
      "Creating sample 42 in mode train at location 1943 on chromosome 1\n",
      "Created sample 42 in 0.0406 seconds\n",
      "Creating sample 43 in mode train at location 1944 on chromosome 1\n",
      "Created sample 43 in 0.0395 seconds\n",
      "Creating sample 44 in mode train at location 2024 on chromosome 1\n",
      "Created sample 44 in 0.0382 seconds\n",
      "Creating sample 45 in mode train at location 2025 on chromosome 1\n",
      "Created sample 45 in 0.0390 seconds\n",
      "Creating sample 46 in mode train at location 2194 on chromosome 1\n",
      "Created sample 46 in 0.0479 seconds\n",
      "Creating sample 47 in mode test at location 2195 on chromosome 1\n",
      "Created sample 47 in 0.0432 seconds\n",
      "Creating sample 48 in mode validation at location 2257 on chromosome 1\n",
      "Created sample 48 in 0.0380 seconds\n",
      "Creating sample 49 in mode train at location 2258 on chromosome 1\n",
      "Created sample 49 in 0.0400 seconds\n",
      "Creating sample 50 in mode test at location 2296 on chromosome 1\n",
      "Created sample 50 in 0.0372 seconds\n",
      "Creating sample 51 in mode train at location 2297 on chromosome 1\n",
      "Created sample 51 in 0.0392 seconds\n",
      "Creating sample 52 in mode train at location 2400 on chromosome 1\n",
      "Created sample 52 in 0.0386 seconds\n",
      "Creating sample 53 in mode validation at location 2401 on chromosome 1\n",
      "Created sample 53 in 0.0390 seconds\n",
      "Creating sample 54 in mode train at location 2431 on chromosome 1\n",
      "Created sample 54 in 0.0374 seconds\n",
      "Creating sample 55 in mode train at location 2432 on chromosome 1\n",
      "Created sample 55 in 0.0401 seconds\n",
      "Creating sample 56 in mode train at location 2437 on chromosome 1\n",
      "Created sample 56 in 0.0387 seconds\n",
      "Creating sample 57 in mode test at location 2438 on chromosome 1\n",
      "Created sample 57 in 0.0390 seconds\n",
      "Creating sample 58 in mode validation at location 2478 on chromosome 1\n",
      "Created sample 58 in 0.0376 seconds\n",
      "Creating sample 59 in mode train at location 2479 on chromosome 1\n",
      "Created sample 59 in 0.0387 seconds\n",
      "Creating sample 60 in mode train at location 2509 on chromosome 1\n",
      "Created sample 60 in 0.0370 seconds\n",
      "Creating sample 61 in mode train at location 2510 on chromosome 1\n",
      "Created sample 61 in 0.0387 seconds\n",
      "Creating sample 62 in mode train at location 2688 on chromosome 1\n",
      "Created sample 62 in 0.0381 seconds\n",
      "Creating sample 63 in mode train at location 2689 on chromosome 1\n",
      "Created sample 63 in 0.0383 seconds\n",
      "Creating sample 64 in mode train at location 2807 on chromosome 1\n",
      "Created sample 64 in 0.0421 seconds\n",
      "Creating sample 65 in mode validation at location 2808 on chromosome 1\n",
      "Created sample 65 in 0.0405 seconds\n",
      "Creating sample 66 in mode train at location 2830 on chromosome 1\n",
      "Created sample 66 in 0.0450 seconds\n",
      "Creating sample 67 in mode train at location 2831 on chromosome 1\n",
      "Created sample 67 in 0.0436 seconds\n",
      "Creating sample 68 in mode train at location 2845 on chromosome 1\n",
      "Created sample 68 in 0.0414 seconds\n",
      "Creating sample 69 in mode train at location 2846 on chromosome 1\n",
      "Created sample 69 in 0.0465 seconds\n",
      "Creating sample 70 in mode train at location 2857 on chromosome 1\n",
      "Created sample 70 in 0.0410 seconds\n",
      "Creating sample 71 in mode train at location 2858 on chromosome 1\n",
      "Created sample 71 in 0.0432 seconds\n",
      "Creating sample 72 in mode train at location 2866 on chromosome 1\n",
      "Created sample 72 in 0.0403 seconds\n",
      "Creating sample 73 in mode train at location 2867 on chromosome 1\n",
      "Created sample 73 in 0.0443 seconds\n",
      "Creating sample 74 in mode test at location 2877 on chromosome 1\n",
      "Created sample 74 in 0.0415 seconds\n",
      "Creating sample 75 in mode train at location 2878 on chromosome 1\n",
      "Created sample 75 in 0.0413 seconds\n",
      "Creating sample 76 in mode train at location 2918 on chromosome 1\n",
      "Created sample 76 in 0.0418 seconds\n",
      "Creating sample 77 in mode validation at location 2919 on chromosome 1\n",
      "Created sample 77 in 0.0407 seconds\n",
      "Creating sample 78 in mode validation at location 2929 on chromosome 1\n",
      "Created sample 78 in 0.0409 seconds\n",
      "Creating sample 79 in mode train at location 2930 on chromosome 1\n",
      "Created sample 79 in 0.0388 seconds\n",
      "Creating sample 80 in mode train at location 2977 on chromosome 1\n",
      "Created sample 80 in 0.0400 seconds\n",
      "Creating sample 81 in mode train at location 2978 on chromosome 1\n",
      "Created sample 81 in 0.0421 seconds\n",
      "Creating sample 82 in mode validation at location 3019 on chromosome 1\n",
      "Created sample 82 in 0.0410 seconds\n",
      "Creating sample 83 in mode train at location 3020 on chromosome 1\n",
      "Created sample 83 in 0.0398 seconds\n",
      "Creating sample 84 in mode train at location 3069 on chromosome 1\n",
      "Created sample 84 in 0.0401 seconds\n",
      "Creating sample 85 in mode validation at location 3070 on chromosome 1\n",
      "Created sample 85 in 0.0431 seconds\n",
      "Creating sample 86 in mode train at location 3091 on chromosome 1\n",
      "Created sample 86 in 0.0385 seconds\n",
      "Creating sample 87 in mode validation at location 3092 on chromosome 1\n",
      "Created sample 87 in 0.0416 seconds\n",
      "Creating sample 88 in mode train at location 3123 on chromosome 1\n",
      "Created sample 88 in 0.0399 seconds\n",
      "Creating sample 89 in mode train at location 3124 on chromosome 1\n",
      "Created sample 89 in 0.0411 seconds\n",
      "Creating sample 90 in mode test at location 3168 on chromosome 1\n",
      "Created sample 90 in 0.0393 seconds\n",
      "Creating sample 91 in mode train at location 3169 on chromosome 1\n",
      "Created sample 91 in 0.0401 seconds\n",
      "Creating sample 92 in mode train at location 3292 on chromosome 1\n",
      "Created sample 92 in 0.0413 seconds\n",
      "Creating sample 93 in mode train at location 3293 on chromosome 1\n",
      "Created sample 93 in 0.0425 seconds\n",
      "Creating sample 94 in mode train at location 3306 on chromosome 1\n",
      "Created sample 94 in 0.0402 seconds\n",
      "Creating sample 95 in mode train at location 3307 on chromosome 1\n",
      "Created sample 95 in 0.0426 seconds\n",
      "Creating sample 96 in mode test at location 3456 on chromosome 1\n",
      "Created sample 96 in 0.0438 seconds\n",
      "Creating sample 97 in mode train at location 3457 on chromosome 1\n",
      "Created sample 97 in 0.0473 seconds\n",
      "Creating sample 98 in mode train at location 3518 on chromosome 1\n",
      "Created sample 98 in 0.0476 seconds\n",
      "Creating sample 99 in mode train at location 3519 on chromosome 1\n",
      "Created sample 99 in 0.0426 seconds\n",
      "Creating sample 100 in mode train at location 3541 on chromosome 1\n",
      "Created sample 100 in 0.0404 seconds\n",
      "Creating sample 101 in mode train at location 3542 on chromosome 1\n",
      "Created sample 101 in 0.0424 seconds\n",
      "Creating sample 102 in mode train at location 3604 on chromosome 1\n",
      "Created sample 102 in 0.0410 seconds\n",
      "Creating sample 103 in mode train at location 3605 on chromosome 1\n",
      "Created sample 103 in 0.0446 seconds\n",
      "Creating sample 104 in mode train at location 3673 on chromosome 1\n",
      "Created sample 104 in 0.0419 seconds\n",
      "Creating sample 105 in mode train at location 3674 on chromosome 1\n",
      "Created sample 105 in 0.0426 seconds\n",
      "Creating sample 106 in mode train at location 3697 on chromosome 1\n",
      "Created sample 106 in 0.0400 seconds\n",
      "Creating sample 107 in mode validation at location 3698 on chromosome 1\n",
      "Created sample 107 in 0.0435 seconds\n",
      "Creating sample 108 in mode train at location 3707 on chromosome 1\n",
      "Created sample 108 in 0.0439 seconds\n",
      "Creating sample 109 in mode train at location 3708 on chromosome 1\n",
      "Created sample 109 in 0.0431 seconds\n",
      "Creating sample 110 in mode test at location 3710 on chromosome 1\n",
      "Created sample 110 in 0.0454 seconds\n",
      "Creating sample 111 in mode train at location 3711 on chromosome 1\n",
      "Created sample 111 in 0.0419 seconds\n",
      "Creating sample 112 in mode train at location 3713 on chromosome 1\n",
      "Created sample 112 in 0.0413 seconds\n",
      "Creating sample 113 in mode train at location 3714 on chromosome 1\n",
      "Created sample 113 in 0.0392 seconds\n",
      "Creating sample 114 in mode train at location 3727 on chromosome 1\n",
      "Created sample 114 in 0.0380 seconds\n",
      "Creating sample 115 in mode train at location 3728 on chromosome 1\n",
      "Created sample 115 in 0.0427 seconds\n",
      "Creating sample 116 in mode train at location 3743 on chromosome 1\n",
      "Created sample 116 in 0.0396 seconds\n",
      "Creating sample 117 in mode train at location 3744 on chromosome 1\n",
      "Created sample 117 in 0.0424 seconds\n",
      "Creating sample 118 in mode train at location 3747 on chromosome 1\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[1;32m/mnt/fast/epigenomics/conschti/model.ipynb Cell 10\u001b[0m line \u001b[0;36m3\n\u001b[1;32m     <a href='vscode-notebook-cell:/mnt/fast/epigenomics/conschti/model.ipynb#X12sZmlsZQ%3D%3D?line=27'>28</a>\u001b[0m         end \u001b[39m=\u001b[39m end \u001b[39m+\u001b[39m \u001b[39m1e6\u001b[39m\n\u001b[1;32m     <a href='vscode-notebook-cell:/mnt/fast/epigenomics/conschti/model.ipynb#X12sZmlsZQ%3D%3D?line=30'>31</a>\u001b[0m tic \u001b[39m=\u001b[39m time\u001b[39m.\u001b[39mperf_counter()\n\u001b[0;32m---> <a href='vscode-notebook-cell:/mnt/fast/epigenomics/conschti/model.ipynb#X12sZmlsZQ%3D%3D?line=31'>32</a>\u001b[0m create_all_samples()\n\u001b[1;32m     <a href='vscode-notebook-cell:/mnt/fast/epigenomics/conschti/model.ipynb#X12sZmlsZQ%3D%3D?line=32'>33</a>\u001b[0m toc \u001b[39m=\u001b[39m time\u001b[39m.\u001b[39mperf_counter()\n\u001b[1;32m     <a href='vscode-notebook-cell:/mnt/fast/epigenomics/conschti/model.ipynb#X12sZmlsZQ%3D%3D?line=33'>34</a>\u001b[0m \u001b[39mprint\u001b[39m(\u001b[39mf\u001b[39m\u001b[39m\"\u001b[39m\u001b[39mCreating all samples in \u001b[39m\u001b[39m{\u001b[39;00mtoc\u001b[39m \u001b[39m\u001b[39m-\u001b[39m\u001b[39m \u001b[39mtic\u001b[39m:\u001b[39;00m\u001b[39m0.4f\u001b[39m\u001b[39m}\u001b[39;00m\u001b[39m seconds\u001b[39m\u001b[39m\"\u001b[39m)\n",
      "\u001b[1;32m/mnt/fast/epigenomics/conschti/model.ipynb Cell 10\u001b[0m line \u001b[0;36m2\n\u001b[1;32m     <a href='vscode-notebook-cell:/mnt/fast/epigenomics/conschti/model.ipynb#X12sZmlsZQ%3D%3D?line=19'>20</a>\u001b[0m mode \u001b[39m=\u001b[39m np\u001b[39m.\u001b[39mrandom\u001b[39m.\u001b[39mchoice([\u001b[39m\"\u001b[39m\u001b[39mtrain\u001b[39m\u001b[39m\"\u001b[39m, \u001b[39m\"\u001b[39m\u001b[39mtest\u001b[39m\u001b[39m\"\u001b[39m, \u001b[39m\"\u001b[39m\u001b[39mvalidation\u001b[39m\u001b[39m\"\u001b[39m], p\u001b[39m=\u001b[39m[\u001b[39m0.8\u001b[39m, \u001b[39m0.1\u001b[39m, \u001b[39m0.1\u001b[39m])\n\u001b[1;32m     <a href='vscode-notebook-cell:/mnt/fast/epigenomics/conschti/model.ipynb#X12sZmlsZQ%3D%3D?line=20'>21</a>\u001b[0m \u001b[39mprint\u001b[39m(\u001b[39mf\u001b[39m\u001b[39m\"\u001b[39m\u001b[39mCreating sample \u001b[39m\u001b[39m{\u001b[39;00mi\u001b[39m}\u001b[39;00m\u001b[39m in mode \u001b[39m\u001b[39m{\u001b[39;00mmode\u001b[39m}\u001b[39;00m\u001b[39m at location \u001b[39m\u001b[39m{\u001b[39;00msite[\u001b[39m'\u001b[39m\u001b[39mlocation\u001b[39m\u001b[39m'\u001b[39m]\u001b[39m}\u001b[39;00m\u001b[39m on chromosome \u001b[39m\u001b[39m{\u001b[39;00msite[\u001b[39m'\u001b[39m\u001b[39mchromosome\u001b[39m\u001b[39m'\u001b[39m]\u001b[39m}\u001b[39;00m\u001b[39m\"\u001b[39m)\n\u001b[0;32m---> <a href='vscode-notebook-cell:/mnt/fast/epigenomics/conschti/model.ipynb#X12sZmlsZQ%3D%3D?line=21'>22</a>\u001b[0m get_neighbours(site, mode)\n\u001b[1;32m     <a href='vscode-notebook-cell:/mnt/fast/epigenomics/conschti/model.ipynb#X12sZmlsZQ%3D%3D?line=22'>23</a>\u001b[0m toc \u001b[39m=\u001b[39m time\u001b[39m.\u001b[39mperf_counter()\n\u001b[1;32m     <a href='vscode-notebook-cell:/mnt/fast/epigenomics/conschti/model.ipynb#X12sZmlsZQ%3D%3D?line=23'>24</a>\u001b[0m \u001b[39mprint\u001b[39m(\u001b[39mf\u001b[39m\u001b[39m\"\u001b[39m\u001b[39mCreated sample \u001b[39m\u001b[39m{\u001b[39;00mi\u001b[39m}\u001b[39;00m\u001b[39m in \u001b[39m\u001b[39m{\u001b[39;00mtoc\u001b[39m \u001b[39m\u001b[39m-\u001b[39m\u001b[39m \u001b[39mtic\u001b[39m:\u001b[39;00m\u001b[39m0.4f\u001b[39m\u001b[39m}\u001b[39;00m\u001b[39m seconds\u001b[39m\u001b[39m\"\u001b[39m)\n",
      "\u001b[1;32m/mnt/fast/epigenomics/conschti/model.ipynb Cell 10\u001b[0m line \u001b[0;36m3\n\u001b[1;32m     <a href='vscode-notebook-cell:/mnt/fast/epigenomics/conschti/model.ipynb#X12sZmlsZQ%3D%3D?line=27'>28</a>\u001b[0m \u001b[39m# print(f\"Histone mods in {toc - tic:0.4f} seconds\")\u001b[39;00m\n\u001b[1;32m     <a href='vscode-notebook-cell:/mnt/fast/epigenomics/conschti/model.ipynb#X12sZmlsZQ%3D%3D?line=28'>29</a>\u001b[0m \u001b[39m# display(h)\u001b[39;00m\n\u001b[1;32m     <a href='vscode-notebook-cell:/mnt/fast/epigenomics/conschti/model.ipynb#X12sZmlsZQ%3D%3D?line=29'>30</a>\u001b[0m \u001b[39m# Get surrounding chromatine states\u001b[39;00m\n\u001b[1;32m     <a href='vscode-notebook-cell:/mnt/fast/epigenomics/conschti/model.ipynb#X12sZmlsZQ%3D%3D?line=30'>31</a>\u001b[0m tic \u001b[39m=\u001b[39m time\u001b[39m.\u001b[39mperf_counter()\n\u001b[0;32m---> <a href='vscode-notebook-cell:/mnt/fast/epigenomics/conschti/model.ipynb#X12sZmlsZQ%3D%3D?line=31'>32</a>\u001b[0m chr_states \u001b[39m=\u001b[39m client\u001b[39m.\u001b[39;49mquery_arrow(\u001b[39mf\u001b[39;49m\u001b[39m\"\u001b[39;49m\u001b[39mselect state, (\u001b[39;49m\u001b[39m{\u001b[39;49;00msite[\u001b[39m'\u001b[39;49m\u001b[39mlocation\u001b[39;49m\u001b[39m'\u001b[39;49m]\u001b[39m}\u001b[39;49;00m\u001b[39m - start ) as start_diff, (\u001b[39;49m\u001b[39m{\u001b[39;49;00msite[\u001b[39m'\u001b[39;49m\u001b[39mlocation\u001b[39;49m\u001b[39m'\u001b[39;49m]\u001b[39m}\u001b[39;49;00m\u001b[39m -end) as end_diff from chr_states where chromosome = \u001b[39;49m\u001b[39m{\u001b[39;49;00msite[\u001b[39m'\u001b[39;49m\u001b[39mchromosome\u001b[39;49m\u001b[39m'\u001b[39;49m]\u001b[39m}\u001b[39;49;00m\u001b[39m order by abs(start_diff) limit \u001b[39;49m\u001b[39m{\u001b[39;49;00mcs_neighbours\u001b[39m \u001b[39;49m\u001b[39m}\u001b[39;49;00m\u001b[39m \u001b[39;49m\u001b[39m\"\u001b[39;49m)\n\u001b[1;32m     <a href='vscode-notebook-cell:/mnt/fast/epigenomics/conschti/model.ipynb#X12sZmlsZQ%3D%3D?line=32'>33</a>\u001b[0m c \u001b[39m=\u001b[39m pl\u001b[39m.\u001b[39mDataFrame(chr_states)\n\u001b[1;32m     <a href='vscode-notebook-cell:/mnt/fast/epigenomics/conschti/model.ipynb#X12sZmlsZQ%3D%3D?line=34'>35</a>\u001b[0m toc \u001b[39m=\u001b[39m time\u001b[39m.\u001b[39mperf_counter()\n",
      "File \u001b[0;32m/mnt/fast/epigenomics/conschti/.env/lib/python3.10/site-packages/clickhouse_connect/driver/client.py:476\u001b[0m, in \u001b[0;36mClient.query_arrow\u001b[0;34m(self, query, parameters, settings, use_strings, external_data)\u001b[0m\n\u001b[1;32m    474\u001b[0m         \u001b[39mraise\u001b[39;00m OperationalError(\u001b[39mf\u001b[39m\u001b[39m'\u001b[39m\u001b[39mCannot change readonly \u001b[39m\u001b[39m{\u001b[39;00marrow_str_setting\u001b[39m}\u001b[39;00m\u001b[39m to \u001b[39m\u001b[39m{\u001b[39;00muse_strings\u001b[39m}\u001b[39;00m\u001b[39m'\u001b[39m)\n\u001b[1;32m    475\u001b[0m     settings[arrow_str_setting] \u001b[39m=\u001b[39m \u001b[39m'\u001b[39m\u001b[39m1\u001b[39m\u001b[39m'\u001b[39m \u001b[39mif\u001b[39;00m use_strings \u001b[39melse\u001b[39;00m \u001b[39m'\u001b[39m\u001b[39m0\u001b[39m\u001b[39m'\u001b[39m\n\u001b[0;32m--> 476\u001b[0m \u001b[39mreturn\u001b[39;00m to_arrow(\u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mraw_query(query,\n\u001b[1;32m    477\u001b[0m                                parameters,\n\u001b[1;32m    478\u001b[0m                                settings,\n\u001b[1;32m    479\u001b[0m                                fmt\u001b[39m=\u001b[39;49m\u001b[39m'\u001b[39;49m\u001b[39mArrow\u001b[39;49m\u001b[39m'\u001b[39;49m,\n\u001b[1;32m    480\u001b[0m                                external_data\u001b[39m=\u001b[39;49mexternal_data))\n",
      "File \u001b[0;32m/mnt/fast/epigenomics/conschti/.env/lib/python3.10/site-packages/clickhouse_connect/driver/httpclient.py:468\u001b[0m, in \u001b[0;36mHttpClient.raw_query\u001b[0;34m(self, query, parameters, settings, fmt, use_database, external_data)\u001b[0m\n\u001b[1;32m    466\u001b[0m     body \u001b[39m=\u001b[39m final_query\n\u001b[1;32m    467\u001b[0m     fields \u001b[39m=\u001b[39m \u001b[39mNone\u001b[39;00m\n\u001b[0;32m--> 468\u001b[0m \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_raw_request(body, params, fields\u001b[39m=\u001b[39;49mfields)\u001b[39m.\u001b[39mdata\n",
      "File \u001b[0;32m/mnt/fast/epigenomics/conschti/.env/lib/python3.10/site-packages/clickhouse_connect/driver/httpclient.py:409\u001b[0m, in \u001b[0;36mHttpClient._raw_request\u001b[0;34m(self, data, params, headers, method, retries, stream, server_wait, fields, error_handler)\u001b[0m\n\u001b[1;32m    407\u001b[0m     \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_active_session \u001b[39m=\u001b[39m query_session\n\u001b[1;32m    408\u001b[0m \u001b[39mtry\u001b[39;00m:\n\u001b[0;32m--> 409\u001b[0m     response \u001b[39m=\u001b[39m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mhttp\u001b[39m.\u001b[39;49mrequest(method, url, \u001b[39m*\u001b[39;49m\u001b[39m*\u001b[39;49mkwargs)\n\u001b[1;32m    410\u001b[0m \u001b[39mexcept\u001b[39;00m HTTPError \u001b[39mas\u001b[39;00m ex:\n\u001b[1;32m    411\u001b[0m     \u001b[39mif\u001b[39;00m \u001b[39misinstance\u001b[39m(ex\u001b[39m.\u001b[39m__context__, \u001b[39mConnectionResetError\u001b[39;00m):\n\u001b[1;32m    412\u001b[0m         \u001b[39m# The server closed the connection, probably because the Keep Alive has expired\u001b[39;00m\n\u001b[1;32m    413\u001b[0m         \u001b[39m# We should be safe to retry, as ClickHouse should not have processed anything on a connection\u001b[39;00m\n\u001b[1;32m    414\u001b[0m         \u001b[39m# that it killed.  We also only retry this once, as multiple disconnects are unlikely to be\u001b[39;00m\n\u001b[1;32m    415\u001b[0m         \u001b[39m# related to the Keep Alive settings\u001b[39;00m\n",
      "File \u001b[0;32m/mnt/fast/epigenomics/conschti/.env/lib/python3.10/site-packages/urllib3/_request_methods.py:118\u001b[0m, in \u001b[0;36mRequestMethods.request\u001b[0;34m(self, method, url, body, fields, headers, json, **urlopen_kw)\u001b[0m\n\u001b[1;32m    110\u001b[0m     \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mrequest_encode_url(\n\u001b[1;32m    111\u001b[0m         method,\n\u001b[1;32m    112\u001b[0m         url,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    115\u001b[0m         \u001b[39m*\u001b[39m\u001b[39m*\u001b[39murlopen_kw,\n\u001b[1;32m    116\u001b[0m     )\n\u001b[1;32m    117\u001b[0m \u001b[39melse\u001b[39;00m:\n\u001b[0;32m--> 118\u001b[0m     \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mrequest_encode_body(\n\u001b[1;32m    119\u001b[0m         method, url, fields\u001b[39m=\u001b[39;49mfields, headers\u001b[39m=\u001b[39;49mheaders, \u001b[39m*\u001b[39;49m\u001b[39m*\u001b[39;49murlopen_kw\n\u001b[1;32m    120\u001b[0m     )\n",
      "File \u001b[0;32m/mnt/fast/epigenomics/conschti/.env/lib/python3.10/site-packages/urllib3/_request_methods.py:217\u001b[0m, in \u001b[0;36mRequestMethods.request_encode_body\u001b[0;34m(self, method, url, fields, headers, encode_multipart, multipart_boundary, **urlopen_kw)\u001b[0m\n\u001b[1;32m    213\u001b[0m     extra_kw[\u001b[39m\"\u001b[39m\u001b[39mheaders\u001b[39m\u001b[39m\"\u001b[39m]\u001b[39m.\u001b[39msetdefault(\u001b[39m\"\u001b[39m\u001b[39mContent-Type\u001b[39m\u001b[39m\"\u001b[39m, content_type)\n\u001b[1;32m    215\u001b[0m extra_kw\u001b[39m.\u001b[39mupdate(urlopen_kw)\n\u001b[0;32m--> 217\u001b[0m \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49murlopen(method, url, \u001b[39m*\u001b[39;49m\u001b[39m*\u001b[39;49mextra_kw)\n",
      "File \u001b[0;32m/mnt/fast/epigenomics/conschti/.env/lib/python3.10/site-packages/urllib3/poolmanager.py:443\u001b[0m, in \u001b[0;36mPoolManager.urlopen\u001b[0;34m(self, method, url, redirect, **kw)\u001b[0m\n\u001b[1;32m    441\u001b[0m     response \u001b[39m=\u001b[39m conn\u001b[39m.\u001b[39murlopen(method, url, \u001b[39m*\u001b[39m\u001b[39m*\u001b[39mkw)\n\u001b[1;32m    442\u001b[0m \u001b[39melse\u001b[39;00m:\n\u001b[0;32m--> 443\u001b[0m     response \u001b[39m=\u001b[39m conn\u001b[39m.\u001b[39;49murlopen(method, u\u001b[39m.\u001b[39;49mrequest_uri, \u001b[39m*\u001b[39;49m\u001b[39m*\u001b[39;49mkw)\n\u001b[1;32m    445\u001b[0m redirect_location \u001b[39m=\u001b[39m redirect \u001b[39mand\u001b[39;00m response\u001b[39m.\u001b[39mget_redirect_location()\n\u001b[1;32m    446\u001b[0m \u001b[39mif\u001b[39;00m \u001b[39mnot\u001b[39;00m redirect_location:\n",
      "File \u001b[0;32m/mnt/fast/epigenomics/conschti/.env/lib/python3.10/site-packages/urllib3/connectionpool.py:791\u001b[0m, in \u001b[0;36mHTTPConnectionPool.urlopen\u001b[0;34m(self, method, url, body, headers, retries, redirect, assert_same_host, timeout, pool_timeout, release_conn, chunked, body_pos, preload_content, decode_content, **response_kw)\u001b[0m\n\u001b[1;32m    788\u001b[0m response_conn \u001b[39m=\u001b[39m conn \u001b[39mif\u001b[39;00m \u001b[39mnot\u001b[39;00m release_conn \u001b[39melse\u001b[39;00m \u001b[39mNone\u001b[39;00m\n\u001b[1;32m    790\u001b[0m \u001b[39m# Make the request on the HTTPConnection object\u001b[39;00m\n\u001b[0;32m--> 791\u001b[0m response \u001b[39m=\u001b[39m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_make_request(\n\u001b[1;32m    792\u001b[0m     conn,\n\u001b[1;32m    793\u001b[0m     method,\n\u001b[1;32m    794\u001b[0m     url,\n\u001b[1;32m    795\u001b[0m     timeout\u001b[39m=\u001b[39;49mtimeout_obj,\n\u001b[1;32m    796\u001b[0m     body\u001b[39m=\u001b[39;49mbody,\n\u001b[1;32m    797\u001b[0m     headers\u001b[39m=\u001b[39;49mheaders,\n\u001b[1;32m    798\u001b[0m     chunked\u001b[39m=\u001b[39;49mchunked,\n\u001b[1;32m    799\u001b[0m     retries\u001b[39m=\u001b[39;49mretries,\n\u001b[1;32m    800\u001b[0m     response_conn\u001b[39m=\u001b[39;49mresponse_conn,\n\u001b[1;32m    801\u001b[0m     preload_content\u001b[39m=\u001b[39;49mpreload_content,\n\u001b[1;32m    802\u001b[0m     decode_content\u001b[39m=\u001b[39;49mdecode_content,\n\u001b[1;32m    803\u001b[0m     \u001b[39m*\u001b[39;49m\u001b[39m*\u001b[39;49mresponse_kw,\n\u001b[1;32m    804\u001b[0m )\n\u001b[1;32m    806\u001b[0m \u001b[39m# Everything went great!\u001b[39;00m\n\u001b[1;32m    807\u001b[0m clean_exit \u001b[39m=\u001b[39m \u001b[39mTrue\u001b[39;00m\n",
      "File \u001b[0;32m/mnt/fast/epigenomics/conschti/.env/lib/python3.10/site-packages/urllib3/connectionpool.py:537\u001b[0m, in \u001b[0;36mHTTPConnectionPool._make_request\u001b[0;34m(self, conn, method, url, body, headers, retries, timeout, chunked, response_conn, preload_content, decode_content, enforce_content_length)\u001b[0m\n\u001b[1;32m    535\u001b[0m \u001b[39m# Receive the response from the server\u001b[39;00m\n\u001b[1;32m    536\u001b[0m \u001b[39mtry\u001b[39;00m:\n\u001b[0;32m--> 537\u001b[0m     response \u001b[39m=\u001b[39m conn\u001b[39m.\u001b[39;49mgetresponse()\n\u001b[1;32m    538\u001b[0m \u001b[39mexcept\u001b[39;00m (BaseSSLError, \u001b[39mOSError\u001b[39;00m) \u001b[39mas\u001b[39;00m e:\n\u001b[1;32m    539\u001b[0m     \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_raise_timeout(err\u001b[39m=\u001b[39me, url\u001b[39m=\u001b[39murl, timeout_value\u001b[39m=\u001b[39mread_timeout)\n",
      "File \u001b[0;32m/mnt/fast/epigenomics/conschti/.env/lib/python3.10/site-packages/urllib3/connection.py:461\u001b[0m, in \u001b[0;36mHTTPConnection.getresponse\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    458\u001b[0m \u001b[39mfrom\u001b[39;00m \u001b[39m.\u001b[39;00m\u001b[39mresponse\u001b[39;00m \u001b[39mimport\u001b[39;00m HTTPResponse\n\u001b[1;32m    460\u001b[0m \u001b[39m# Get the response from http.client.HTTPConnection\u001b[39;00m\n\u001b[0;32m--> 461\u001b[0m httplib_response \u001b[39m=\u001b[39m \u001b[39msuper\u001b[39;49m()\u001b[39m.\u001b[39;49mgetresponse()\n\u001b[1;32m    463\u001b[0m \u001b[39mtry\u001b[39;00m:\n\u001b[1;32m    464\u001b[0m     assert_header_parsing(httplib_response\u001b[39m.\u001b[39mmsg)\n",
      "File \u001b[0;32m/usr/lib/python3.10/http/client.py:1375\u001b[0m, in \u001b[0;36mHTTPConnection.getresponse\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m   1373\u001b[0m \u001b[39mtry\u001b[39;00m:\n\u001b[1;32m   1374\u001b[0m     \u001b[39mtry\u001b[39;00m:\n\u001b[0;32m-> 1375\u001b[0m         response\u001b[39m.\u001b[39;49mbegin()\n\u001b[1;32m   1376\u001b[0m     \u001b[39mexcept\u001b[39;00m \u001b[39mConnectionError\u001b[39;00m:\n\u001b[1;32m   1377\u001b[0m         \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mclose()\n",
      "File \u001b[0;32m/usr/lib/python3.10/http/client.py:318\u001b[0m, in \u001b[0;36mHTTPResponse.begin\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    316\u001b[0m \u001b[39m# read until we get a non-100 response\u001b[39;00m\n\u001b[1;32m    317\u001b[0m \u001b[39mwhile\u001b[39;00m \u001b[39mTrue\u001b[39;00m:\n\u001b[0;32m--> 318\u001b[0m     version, status, reason \u001b[39m=\u001b[39m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_read_status()\n\u001b[1;32m    319\u001b[0m     \u001b[39mif\u001b[39;00m status \u001b[39m!=\u001b[39m CONTINUE:\n\u001b[1;32m    320\u001b[0m         \u001b[39mbreak\u001b[39;00m\n",
      "File \u001b[0;32m/usr/lib/python3.10/http/client.py:279\u001b[0m, in \u001b[0;36mHTTPResponse._read_status\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    278\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39m_read_status\u001b[39m(\u001b[39mself\u001b[39m):\n\u001b[0;32m--> 279\u001b[0m     line \u001b[39m=\u001b[39m \u001b[39mstr\u001b[39m(\u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mfp\u001b[39m.\u001b[39;49mreadline(_MAXLINE \u001b[39m+\u001b[39;49m \u001b[39m1\u001b[39;49m), \u001b[39m\"\u001b[39m\u001b[39miso-8859-1\u001b[39m\u001b[39m\"\u001b[39m)\n\u001b[1;32m    280\u001b[0m     \u001b[39mif\u001b[39;00m \u001b[39mlen\u001b[39m(line) \u001b[39m>\u001b[39m _MAXLINE:\n\u001b[1;32m    281\u001b[0m         \u001b[39mraise\u001b[39;00m LineTooLong(\u001b[39m\"\u001b[39m\u001b[39mstatus line\u001b[39m\u001b[39m\"\u001b[39m)\n",
      "File \u001b[0;32m/usr/lib/python3.10/socket.py:705\u001b[0m, in \u001b[0;36mSocketIO.readinto\u001b[0;34m(self, b)\u001b[0m\n\u001b[1;32m    703\u001b[0m \u001b[39mwhile\u001b[39;00m \u001b[39mTrue\u001b[39;00m:\n\u001b[1;32m    704\u001b[0m     \u001b[39mtry\u001b[39;00m:\n\u001b[0;32m--> 705\u001b[0m         \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_sock\u001b[39m.\u001b[39;49mrecv_into(b)\n\u001b[1;32m    706\u001b[0m     \u001b[39mexcept\u001b[39;00m timeout:\n\u001b[1;32m    707\u001b[0m         \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_timeout_occurred \u001b[39m=\u001b[39m \u001b[39mTrue\u001b[39;00m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "import time\n",
    "import numpy as np\n",
    "\n",
    "\n",
    "\n",
    "def create_all_samples():\n",
    "\n",
    "    start = 0\n",
    "    end = 1e6\n",
    "    total = 557173708\n",
    "\n",
    "    while start < total:\n",
    "        next_samples = client.query_arrow(f\"select * except (trinucleotide_context, pedigree, id) from methylome where generation = 0 and context = 'CG' and location between {start} and {end} order by chromosome, location, strand\")\n",
    "        df = pl.DataFrame(next_samples)\n",
    "\n",
    "        print(f\"Creating {df.height} samples in range {start} to {end}\")\n",
    "        i = 0\n",
    "        for site in df.iter_rows(named=True):\n",
    "            tic = time.perf_counter()\n",
    "            mode = np.random.choice([\"train\", \"test\", \"validation\"], p=[0.8, 0.1, 0.1])\n",
    "            print(f\"Creating sample {i} in mode {mode} at location {site['location']} on chromosome {site['chromosome']}\")\n",
    "            get_neighbours(site, mode)\n",
    "            toc = time.perf_counter()\n",
    "            print(f\"Created sample {i} in {toc - tic:0.4f} seconds\")\n",
    "            i += 1\n",
    "        \n",
    "        start = end\n",
    "        end = end + 1e6\n",
    "\n",
    "\n",
    "tic = time.perf_counter()\n",
    "# create_all_samples()\n",
    "toc = time.perf_counter()\n",
    "print(f\"Creating all samples in {toc - tic:0.4f} seconds\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def get_batch(mode):\n",
    "    offset = 0. if mode == \"train\" else 0.9\n",
    "    sample = 0.9 if mode == \"train\" else 0.1\n",
    "    tic = time.perf_counter()\n",
    "    sites =  client.query_arrow(f\"select * except (trinucleotide_context, pedigree, id) from methylome sample {sample} offset {offset} where generation = 0 order by rand() limit {batch_size}\")\n",
    "    toc = time.perf_counter()\n",
    "\n",
    "    #print(f\"fetch batch in {toc - tic:0.4f} seconds\")\n",
    "    df = pl.DataFrame(sites)\n",
    "    # display(df)\n",
    "\n",
    "    xs = []\n",
    "    targets = []\n",
    "    for site in df.iter_rows(named=True):\n",
    "        x, target = get_neighbours(site)\n",
    "        xs.append(x)\n",
    "        targets.append(target)\n",
    "   \n",
    "    data = torch.cat(xs).to(device)\n",
    "    targets = torch.cat(targets).to(device)\n",
    "    return data, targets\n",
    "    \n",
    "tic = time.perf_counter()\n",
    "data, targets = get_batch(\"train\")\n",
    "toc = time.perf_counter()\n",
    "print(f\"Everything in {toc - tic:0.4f} seconds\")\n",
    "print(data, targets)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Defining the model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "from torch import nn\n",
    "from torch.nn import functional as F\n",
    "\n",
    "class MethylationMaster(nn.Module):\n",
    "    def __init__(self):\n",
    "        super().__init__()\n",
    "\n",
    "        self.transformer = nn.Transformer(d_model=330, nhead=33, dtype=torch.float32)\n",
    "\n",
    "    def forward(self, x, c, targets = None):\n",
    "        logits =  self.transformer(c, x, tgt_is_causal=True)\n",
    "        print(logits.shape) # Should be (gene_length * 6)\n",
    "\n",
    "        if targets is None:\n",
    "            return logits, None\n",
    "\n",
    "        loss = F.mse_loss(logits, targets)\n",
    "\n",
    "        return logits, loss\n",
    "    \n",
    "    @torch.no_grad()\n",
    "    def generate(self, idx):\n",
    "        logits = self.transformer(idx, idx) # (B * S, C)\n",
    "\n",
    "        last = logits[-1]\n",
    "        last_tar = targets[-1]\n",
    "\n",
    "        probs = F.softmax(last, dim=-1) \n",
    "        # sample from the distribution\n",
    "        guess = torch.multinomial(probs, num_samples=1) # (B, 1)\n",
    "\n",
    "\n",
    "        print(f\"Guess: {guess}, correct: {last_tar}\")\n",
    "\n",
    "        return guess == last_tar"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Training"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "24.134376 M parameters\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/mnt/fast/epigenomics/conschti/.env/lib/python3.10/site-packages/torch/nn/modules/transformer.py:282: UserWarning: enable_nested_tensor is True, but self.use_nested_tensor is False because encoder_layer.self_attn.batch_first was not True(use batch_first for better inference performance)\n",
      "  warnings.warn(f\"enable_nested_tensor is True, but self.use_nested_tensor is False because {why_not_sparsity_fast_path}\")\n"
     ]
    }
   ],
   "source": [
    "from torch.utils.data import DataLoader\n",
    "\n",
    "training_data = Sites(mode=\"train\")\n",
    "train_dataloader = DataLoader(training_data, batch_size=None)\n",
    "conschti = MethylationMaster().to(device)\n",
    "\n",
    "print(sum(p.numel() for p in conschti.parameters())/1e6, 'M parameters')\n",
    "optimizer = torch.optim.Adam(conschti.parameters(), lr=1e-3)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "ename": "RuntimeError",
     "evalue": "shape '[2065, 6]' is invalid for input of size 0",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mRuntimeError\u001b[0m                              Traceback (most recent call last)",
      "\u001b[1;32m/mnt/fast/epigenomics/conschti/model.ipynb Cell 14\u001b[0m line \u001b[0;36m6\n\u001b[1;32m      <a href='vscode-notebook-cell:/mnt/fast/epigenomics/conschti/model.ipynb#X16sZmlsZQ%3D%3D?line=2'>3</a>\u001b[0m \u001b[39mfor\u001b[39;00m i \u001b[39min\u001b[39;00m \u001b[39mrange\u001b[39m(trainings_steps):\n\u001b[1;32m      <a href='vscode-notebook-cell:/mnt/fast/epigenomics/conschti/model.ipynb#X16sZmlsZQ%3D%3D?line=3'>4</a>\u001b[0m     optimizer\u001b[39m.\u001b[39mzero_grad()\n\u001b[0;32m----> <a href='vscode-notebook-cell:/mnt/fast/epigenomics/conschti/model.ipynb#X16sZmlsZQ%3D%3D?line=5'>6</a>\u001b[0m     x, c, t \u001b[39m=\u001b[39m \u001b[39mnext\u001b[39;49m(\u001b[39miter\u001b[39;49m(train_dataloader))\n\u001b[1;32m      <a href='vscode-notebook-cell:/mnt/fast/epigenomics/conschti/model.ipynb#X16sZmlsZQ%3D%3D?line=8'>9</a>\u001b[0m     logits, loss \u001b[39m=\u001b[39m conschti(x, c , t)\n\u001b[1;32m     <a href='vscode-notebook-cell:/mnt/fast/epigenomics/conschti/model.ipynb#X16sZmlsZQ%3D%3D?line=9'>10</a>\u001b[0m     loss\u001b[39m.\u001b[39mbackward()\n",
      "File \u001b[0;32m/mnt/fast/epigenomics/conschti/.env/lib/python3.10/site-packages/torch/utils/data/dataloader.py:630\u001b[0m, in \u001b[0;36m_BaseDataLoaderIter.__next__\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    627\u001b[0m \u001b[39mif\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_sampler_iter \u001b[39mis\u001b[39;00m \u001b[39mNone\u001b[39;00m:\n\u001b[1;32m    628\u001b[0m     \u001b[39m# TODO(https://github.com/pytorch/pytorch/issues/76750)\u001b[39;00m\n\u001b[1;32m    629\u001b[0m     \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_reset()  \u001b[39m# type: ignore[call-arg]\u001b[39;00m\n\u001b[0;32m--> 630\u001b[0m data \u001b[39m=\u001b[39m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_next_data()\n\u001b[1;32m    631\u001b[0m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_num_yielded \u001b[39m+\u001b[39m\u001b[39m=\u001b[39m \u001b[39m1\u001b[39m\n\u001b[1;32m    632\u001b[0m \u001b[39mif\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_dataset_kind \u001b[39m==\u001b[39m _DatasetKind\u001b[39m.\u001b[39mIterable \u001b[39mand\u001b[39;00m \\\n\u001b[1;32m    633\u001b[0m         \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_IterableDataset_len_called \u001b[39mis\u001b[39;00m \u001b[39mnot\u001b[39;00m \u001b[39mNone\u001b[39;00m \u001b[39mand\u001b[39;00m \\\n\u001b[1;32m    634\u001b[0m         \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_num_yielded \u001b[39m>\u001b[39m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_IterableDataset_len_called:\n",
      "File \u001b[0;32m/mnt/fast/epigenomics/conschti/.env/lib/python3.10/site-packages/torch/utils/data/dataloader.py:674\u001b[0m, in \u001b[0;36m_SingleProcessDataLoaderIter._next_data\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    672\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39m_next_data\u001b[39m(\u001b[39mself\u001b[39m):\n\u001b[1;32m    673\u001b[0m     index \u001b[39m=\u001b[39m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_next_index()  \u001b[39m# may raise StopIteration\u001b[39;00m\n\u001b[0;32m--> 674\u001b[0m     data \u001b[39m=\u001b[39m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_dataset_fetcher\u001b[39m.\u001b[39;49mfetch(index)  \u001b[39m# may raise StopIteration\u001b[39;00m\n\u001b[1;32m    675\u001b[0m     \u001b[39mif\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_pin_memory:\n\u001b[1;32m    676\u001b[0m         data \u001b[39m=\u001b[39m _utils\u001b[39m.\u001b[39mpin_memory\u001b[39m.\u001b[39mpin_memory(data, \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_pin_memory_device)\n",
      "File \u001b[0;32m/mnt/fast/epigenomics/conschti/.env/lib/python3.10/site-packages/torch/utils/data/_utils/fetch.py:41\u001b[0m, in \u001b[0;36m_IterableDatasetFetcher.fetch\u001b[0;34m(self, possibly_batched_index)\u001b[0m\n\u001b[1;32m     39\u001b[0m         \u001b[39mraise\u001b[39;00m \u001b[39mStopIteration\u001b[39;00m\n\u001b[1;32m     40\u001b[0m \u001b[39melse\u001b[39;00m:\n\u001b[0;32m---> 41\u001b[0m     data \u001b[39m=\u001b[39m \u001b[39mnext\u001b[39;49m(\u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mdataset_iter)\n\u001b[1;32m     42\u001b[0m \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mcollate_fn(data)\n",
      "\u001b[1;32m/mnt/fast/epigenomics/conschti/model.ipynb Cell 14\u001b[0m line \u001b[0;36m4\n\u001b[1;32m     <a href='vscode-notebook-cell:/mnt/fast/epigenomics/conschti/model.ipynb#X16sZmlsZQ%3D%3D?line=44'>45</a>\u001b[0m targets_line_2 \u001b[39m=\u001b[39m torch\u001b[39m.\u001b[39mtensor(targets_line_2, dtype\u001b[39m=\u001b[39mtorch\u001b[39m.\u001b[39mfloat32)\n\u001b[1;32m     <a href='vscode-notebook-cell:/mnt/fast/epigenomics/conschti/model.ipynb#X16sZmlsZQ%3D%3D?line=45'>46</a>\u001b[0m \u001b[39m# targets_line_8 = torch.tensor(targets_line_8, dtype=torch.float32)\u001b[39;00m\n\u001b[0;32m---> <a href='vscode-notebook-cell:/mnt/fast/epigenomics/conschti/model.ipynb#X16sZmlsZQ%3D%3D?line=47'>48</a>\u001b[0m targets_line_2 \u001b[39m=\u001b[39m targets_line_2\u001b[39m.\u001b[39;49mreshape(gene_length, \u001b[39m6\u001b[39;49m)\n\u001b[1;32m     <a href='vscode-notebook-cell:/mnt/fast/epigenomics/conschti/model.ipynb#X16sZmlsZQ%3D%3D?line=48'>49</a>\u001b[0m \u001b[39m# targets_line_8 = targets_line_8.reshape(gene_length, 6)\u001b[39;00m\n\u001b[1;32m     <a href='vscode-notebook-cell:/mnt/fast/epigenomics/conschti/model.ipynb#X16sZmlsZQ%3D%3D?line=50'>51</a>\u001b[0m targets_line_2 \u001b[39m=\u001b[39m targets_line_2[\u001b[39m1\u001b[39m:, :] \u001b[39m# remove first row, these can not be predicted as there is no predecessor\u001b[39;00m\n",
      "\u001b[0;31mRuntimeError\u001b[0m: shape '[2065, 6]' is invalid for input of size 0"
     ]
    }
   ],
   "source": [
    "\n",
    "trainings_steps = 1000\n",
    "\n",
    "for i in range(trainings_steps):\n",
    "    optimizer.zero_grad()\n",
    "\n",
    "    x, c, t = next(iter(train_dataloader))\n",
    "\n",
    "\n",
    "    logits, loss = conschti(x, c , t)\n",
    "    loss.backward()\n",
    "    optimizer.step()\n",
    "    \n",
    "    print(f\"Step {i}, loss: {loss.item()}\")\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
