{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "def query(gene_type = 'gbM', modification = 'H2AZ'):\n",
    "\n",
    "    return f\"\"\"\n",
    "    with genes_chr1 as (Select start, \"end\", strand, chromosome from genes where type = '{gene_type}' and chromosome = 1),\n",
    "genes_chr2 as (Select start, \"end\", strand, chromosome from genes where type = '{gene_type}' and chromosome = 2),\n",
    "genes_chr3 as (Select start, \"end\", strand, chromosome from genes where type = '{gene_type}' and chromosome = 3),\n",
    "genes_chr4 as (Select start, \"end\", strand, chromosome from genes where type = '{gene_type}' and chromosome = 4),\n",
    "genes_chr5 as (Select start, \"end\", strand, chromosome from genes where type = '{gene_type}' and chromosome = 5),\n",
    "\n",
    "percentiles_chr1 as (\n",
    "select Count(*) as sites,\n",
    "        FLOOR(\n",
    "            (\n",
    "                CASE\n",
    "                    WHEN a.strand = '+' then (c.start - a.start)\n",
    "                    else (a.end - c.\"end\"\n",
    "                )\n",
    "            end * 100\n",
    "        ) / (a.end - a.start\n",
    ")\n",
    ") as percentile\n",
    "from histone_mods c, genes_chr1 a WHERE \n",
    "    c.chromosome = 1\n",
    "    and c.modification = '{modification}'\n",
    "and c.start + (a.end-a.start) >= a.start\n",
    "and c.\"end\" - (a.end -a.start) <= a.\"end\"\n",
    "group by percentile order by percentile),\n",
    "\n",
    "percentiles_chr2 as (\n",
    "select Count(*) as sites,\n",
    "        FLOOR(\n",
    "            (\n",
    "                CASE\n",
    "                    WHEN a.strand = '+' then (c.start - a.start)\n",
    "                    else (a.end - c.\"end\"\n",
    "                )\n",
    "            end * 100\n",
    "        ) / (a.end - a.start\n",
    ")\n",
    ") as percentile\n",
    "from histone_mods c, genes_chr2 a WHERE \n",
    "    c.chromosome = 2\n",
    "    and c.modification = '{modification}'\n",
    "and c.start + (a.end-a.start) >= a.start\n",
    "and c.\"end\" - (a.end -a.start) <= a.\"end\"\n",
    "group by percentile order by percentile),\n",
    "\n",
    "percentiles_chr3 as(\n",
    "select Count(*) as sites,\n",
    "        FLOOR(\n",
    "            (\n",
    "                CASE\n",
    "                    WHEN a.strand = '+' then (c.start - a.start)\n",
    "                    else (a.end - c.\"end\"\n",
    "                )\n",
    "            end * 100\n",
    "        ) / (a.end - a.start\n",
    ")\n",
    ") as percentile\n",
    "from histone_mods c, genes_chr3 a WHERE \n",
    "    c.chromosome = 3\n",
    "    and c.modification = '{modification}'\n",
    "and c.start + (a.end-a.start) >= a.start\n",
    "and c.\"end\" - (a.end -a.start) <= a.\"end\"\n",
    "group by percentile order by percentile),\n",
    "\n",
    "percentiles_chr4 as (\n",
    "select Count(*) as sites,\n",
    "        FLOOR(\n",
    "            (\n",
    "                CASE\n",
    "                    WHEN a.strand = '+' then (c.start - a.start)\n",
    "                    else (a.end - c.\"end\"\n",
    "                )\n",
    "            end * 100\n",
    "        ) / (a.end - a.start\n",
    ")\n",
    ") as percentile\n",
    "from histone_mods c, genes_chr4 a WHERE \n",
    "    c.chromosome = 4\n",
    "    and c.modification = '{modification}'\n",
    "and c.start + (a.end-a.start) >= a.start\n",
    "and c.\"end\" - (a.end -a.start) <= a.\"end\"\n",
    "group by percentile order by percentile),\n",
    "\n",
    "percentiles_chr5 as (\n",
    "select Count(*) as sites,\n",
    "        FLOOR(\n",
    "            (\n",
    "                CASE\n",
    "                    WHEN a.strand = '+' then (c.start - a.start)\n",
    "                    else (a.end - c.\"end\"\n",
    "                )\n",
    "            end * 100\n",
    "        ) / (a.end - a.start\n",
    ")\n",
    ") as percentile\n",
    "from histone_mods c, genes_chr5 a WHERE \n",
    "    c.chromosome = 5\n",
    "    and c.modification = '{modification}'\n",
    "and c.start + (a.end-a.start) >= a.start\n",
    "and c.\"end\" - (a.end -a.start) <= a.\"end\"\n",
    "group by percentile order by percentile)\n",
    "\n",
    "select percentiles_chr1.sites + percentiles_chr2.sites + percentiles_chr3.sites + percentiles_chr4.sites + percentiles_chr5.sites as sites, percentiles_chr1.percentile as percentile\n",
    "from percentiles_chr1, percentiles_chr2, percentiles_chr3, percentiles_chr4, percentiles_chr5\n",
    "where  \n",
    "percentiles_chr1.percentile = percentiles_chr2.percentile and\n",
    "percentiles_chr1.percentile = percentiles_chr3.percentile and\n",
    "percentiles_chr1.percentile = percentiles_chr4.percentile and\n",
    "percentiles_chr1.percentile = percentiles_chr5.percentile\n",
    "order by percentile\n",
    "    \"\"\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'\\n    with genes_chr1 as (Select start, \"end\", strand, chromosome from genes where type = UM and chromosome = 1),\\ngenes_chr2 as (Select start, \"end\", strand, chromosome from genes where type = UM and chromosome = 2),\\ngenes_chr3 as (Select start, \"end\", strand, chromosome from genes where type = UM and chromosome = 3),\\ngenes_chr4 as (Select start, \"end\", strand, chromosome from genes where type = UM and chromosome = 4),\\ngenes_chr5 as (Select start, \"end\", strand, chromosome from genes where type = UM and chromosome = 5),\\n\\npercentiles_chr1 as (\\nselect Count(*) as sites,\\n        FLOOR(\\n            (\\n                CASE\\n                    WHEN a.strand = \\'+\\' then (c.start - a.start)\\n                    else (a.end - c.\"end\"\\n                )\\n            end * 100\\n        ) / (a.end - a.start\\n)\\n) as percentile\\nfrom histone_mods c, genes_chr1 a WHERE \\n    c.chromosome = 1\\n    and c.modification = H3K27Me3\\nand c.start + (a.end-a.start) >= a.start\\nand c.\"end\" - (a.end -a.start) <= a.\"end\"\\ngroup by percentile order by percentile),\\n\\npercentiles_chr2 as (\\nselect Count(*) as sites,\\n        FLOOR(\\n            (\\n                CASE\\n                    WHEN a.strand = \\'+\\' then (c.start - a.start)\\n                    else (a.end - c.\"end\"\\n                )\\n            end * 100\\n        ) / (a.end - a.start\\n)\\n) as percentile\\nfrom histone_mods c, genes_chr2 a WHERE \\n    c.chromosome = 2\\n    and c.modification = H3K27Me3\\nand c.start + (a.end-a.start) >= a.start\\nand c.\"end\" - (a.end -a.start) <= a.\"end\"\\ngroup by percentile order by percentile),\\n\\npercentiles_chr3 as(\\nselect Count(*) as sites,\\n        FLOOR(\\n            (\\n                CASE\\n                    WHEN a.strand = \\'+\\' then (c.start - a.start)\\n                    else (a.end - c.\"end\"\\n                )\\n            end * 100\\n        ) / (a.end - a.start\\n)\\n) as percentile\\nfrom histone_mods c, genes_chr3 a WHERE \\n    c.chromosome = 3\\n    and c.modification = H3K27Me3\\nand c.start + (a.end-a.start) >= a.start\\nand c.\"end\" - (a.end -a.start) <= a.\"end\"\\ngroup by percentile order by percentile),\\n\\npercentiles_chr4 as (\\nselect Count(*) as sites,\\n        FLOOR(\\n            (\\n                CASE\\n                    WHEN a.strand = \\'+\\' then (c.start - a.start)\\n                    else (a.end - c.\"end\"\\n                )\\n            end * 100\\n        ) / (a.end - a.start\\n)\\n) as percentile\\nfrom histone_mods c, genes_chr4 a WHERE \\n    c.chromosome = 4\\n    and c.modification = H3K27Me3\\nand c.start + (a.end-a.start) >= a.start\\nand c.\"end\" - (a.end -a.start) <= a.\"end\"\\ngroup by percentile order by percentile),\\n\\npercentiles_chr5 as (\\nselect Count(*) as sites,\\n        FLOOR(\\n            (\\n                CASE\\n                    WHEN a.strand = \\'+\\' then (c.start - a.start)\\n                    else (a.end - c.\"end\"\\n                )\\n            end * 100\\n        ) / (a.end - a.start\\n)\\n) as percentile\\nfrom histone_mods c, genes_chr5 a WHERE \\n    c.chromosome = 5\\n    and c.modification = H3K27Me3\\nand c.start + (a.end-a.start) >= a.start\\nand c.\"end\" - (a.end -a.start) <= a.\"end\"\\ngroup by percentile order by percentile)\\n\\nselect percentiles_chr1.sites + percentiles_chr2.sites + percentiles_chr3.sites + percentiles_chr4.sites + percentiles_chr5.sites as sites, percentiles_chr1.percentile as percentile\\nfrom percentiles_chr1, percentiles_chr2, percentiles_chr3, percentiles_chr4, percentiles_chr5\\nwhere  \\npercentiles_chr1.percentile = percentiles_chr2.percentile and\\npercentiles_chr1.percentile = percentiles_chr3.percentile and\\npercentiles_chr1.percentile = percentiles_chr4.percentile and\\npercentiles_chr1.percentile = percentiles_chr5.percentile\\norder by percentile;\\n    '"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query(\"UM\", \"H3K27Me3\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import clickhouse_connect\n",
    "import polars as pl\n",
    "client = clickhouse_connect.get_client(host='localhost', username='cgoeldel', password='Goe1409ldel')\n",
    "\n",
    "results = {}\n",
    "\n",
    "for gene_type in [\"UM\", \"teM\", \"gbM\"]:\n",
    "    for modification in [\"H2AZ\",\"H3\",\"H3K36Me3\",\"H3K4Me1\",\"H3K27Me3\",\"H3K4Me3\",\"H3K56Ac\",\"input\"]:\n",
    "        res = client.query_arrow(query(gene_type, modification))\n",
    "        df = pl.DataFrame(res)\n",
    "        df.write_csv(f\"/mnt/fast/epigenomics/modifications/percentiles/{gene_type}_{modification}.csv\")\n",
    "        results[f\"{gene_type}_{modification}\"] = df\n",
    "        \n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "def chr_state_query(gene_type='gbM', state = 1):\n",
    "    return f\"\"\"\n",
    "    with genes_chr1 as (Select start, \"end\", strand, chromosome from genes where type = '{gene_type}' and chromosome = 1),\n",
    "genes_chr2 as (Select start, \"end\", strand, chromosome from genes where type = '{gene_type}' and chromosome = 2),\n",
    "genes_chr3 as (Select start, \"end\", strand, chromosome from genes where type = '{gene_type}' and chromosome = 3),\n",
    "genes_chr4 as (Select start, \"end\", strand, chromosome from genes where type = '{gene_type}' and chromosome = 4),\n",
    "genes_chr5 as (Select start, \"end\", strand, chromosome from genes where type = '{gene_type}' and chromosome = 5),\n",
    "\n",
    "percentiles_chr1 as (\n",
    "select Count(*) as sites,\n",
    "        FLOOR(\n",
    "            (\n",
    "                CASE\n",
    "                    WHEN a.strand = '+' then ((c.start + c.end) / 2 - a.start)\n",
    "                    ELSE (a.end - (c.start + c.\"end\") / 2) END * 100\n",
    "        ) / (a.end - a.start)\n",
    ") as percentile\n",
    "from chr_states c, genes_chr1 a WHERE \n",
    "    c.chromosome = 1\n",
    "    and c.state = '{state}'\n",
    "and (c.start + c.end) / 2 + (a.end- a.start) >= a.start\n",
    "and (c.start + c.end) / 2 - (a.end -a.start) <= a.\"end\"\n",
    "group by percentile order by percentile),\n",
    "\n",
    "percentiles_chr2 as (\n",
    "select Count(*) as sites,\n",
    "        FLOOR(\n",
    "            (\n",
    "                CASE\n",
    "                    WHEN a.strand = '+' then ((c.start + c.end) / 2 - a.start)\n",
    "                    ELSE (a.end - (c.start + c.\"end\") / 2) END * 100\n",
    "        ) / (a.end - a.start)\n",
    ") as percentile\n",
    "from chr_states c, genes_chr2 a WHERE \n",
    "    c.chromosome = 2\n",
    "    and c.state = '{state}'\n",
    "and (c.start + c.end) / 2 + (a.end- a.start) >= a.start\n",
    "and (c.start + c.end) / 2 - (a.end -a.start) <= a.\"end\"\n",
    "group by percentile order by percentile),\n",
    "\n",
    "percentiles_chr3 as(\n",
    "select Count(*) as sites,\n",
    "        FLOOR(\n",
    "            (\n",
    "                CASE\n",
    "                    WHEN a.strand = '+' then ((c.start + c.end) / 2 - a.start)\n",
    "                    ELSE (a.end - (c.start + c.\"end\") / 2) END * 100\n",
    "        ) / (a.end - a.start)\n",
    ") as percentile\n",
    "from chr_states c, genes_chr3 a WHERE \n",
    "    c.chromosome = 3\n",
    "    and c.state = '{state}'\n",
    "and (c.start + c.end) / 2 + (a.end- a.start) >= a.start\n",
    "and (c.start + c.end) / 2 - (a.end -a.start) <= a.\"end\"\n",
    "group by percentile order by percentile),\n",
    "\n",
    "percentiles_chr4 as (\n",
    "select Count(*) as sites,\n",
    "        FLOOR(\n",
    "            (\n",
    "                CASE\n",
    "                    WHEN a.strand = '+' then ((c.start + c.end) / 2 - a.start)\n",
    "                    ELSE (a.end - (c.start + c.\"end\") / 2) END * 100\n",
    "        ) / (a.end - a.start)\n",
    ") as percentile\n",
    "from chr_states c, genes_chr4 a WHERE \n",
    "    c.chromosome = 4\n",
    "    and c.state = '{state}'\n",
    "and (c.start + c.end) / 2 + (a.end- a.start) >= a.start\n",
    "and (c.start + c.end) / 2 - (a.end -a.start) <= a.\"end\"\n",
    "group by percentile order by percentile),\n",
    "\n",
    "percentiles_chr5 as (\n",
    "select Count(*) as sites,\n",
    "        FLOOR(\n",
    "            (\n",
    "                CASE\n",
    "                    WHEN a.strand = '+' then ((c.start + c.end) / 2 - a.start)\n",
    "                    ELSE (a.end - (c.start + c.\"end\") / 2) END * 100\n",
    "        ) / (a.end - a.start)\n",
    ") as percentile\n",
    "from chr_states c, genes_chr5 a WHERE \n",
    "    c.chromosome = 5\n",
    "    and c.state = '{state}'\n",
    "and (c.start + c.end) / 2 + (a.end- a.start) >= a.start\n",
    "and (c.start + c.end) / 2 - (a.end -a.start) <= a.\"end\"\n",
    "group by percentile order by percentile)\n",
    "\n",
    "select percentiles_chr1.sites + percentiles_chr2.sites + percentiles_chr3.sites + percentiles_chr4.sites + percentiles_chr5.sites as sites, percentiles_chr1.percentile as percentile\n",
    "from percentiles_chr1, percentiles_chr2, percentiles_chr3, percentiles_chr4, percentiles_chr5\n",
    "where  \n",
    "percentiles_chr1.percentile = percentiles_chr2.percentile and\n",
    "percentiles_chr1.percentile = percentiles_chr3.percentile and\n",
    "percentiles_chr1.percentile = percentiles_chr4.percentile and\n",
    "percentiles_chr1.percentile = percentiles_chr5.percentile\n",
    "order by percentile\n",
    "    \"\"\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def meth_sites_query(gene_type='gbM', context = 'CG'):\n",
    "    return f\"\"\"\n",
    "with genes_chr1 as (Select start, \"end\", strand, chromosome from genes where type = '{gene_type}' and chromosome = 1),\n",
    "genes_chr2 as (Select start, \"end\", strand, chromosome from genes where type = '{gene_type}' and chromosome = 2),\n",
    "genes_chr3 as (Select start, \"end\", strand, chromosome from genes where type = '{gene_type}' and chromosome = 3),\n",
    "genes_chr4 as (Select start, \"end\", strand, chromosome from genes where type = '{gene_type}' and chromosome = 4),\n",
    "genes_chr5 as (Select start, \"end\", strand, chromosome from genes where type = '{gene_type}' and chromosome = 5),\n",
    "sites_chr1 as (Select * from methylome where context = '{context}' and generation = 0 and chromosome = 1),\n",
    "sites_chr2 as (Select * from methylome where context = '{context}' and generation = 0 and chromosome = 2),\n",
    "sites_chr3 as (Select * from methylome where context = '{context}' and generation = 0 and chromosome = 3),\n",
    "sites_chr4 as (Select * from methylome where context = '{context}' and generation = 0 and chromosome = 4),\n",
    "sites_chr5 as (Select * from methylome where context = '{context}' and generation = 0 and chromosome = 5),\n",
    "\n",
    "percentiles_chr1 as (\n",
    "select Count(*) as sites,\n",
    "        FLOOR(\n",
    "            (\n",
    "                CASE\n",
    "                    WHEN a.strand = '+' then (s.location - a.start)\n",
    "                    ELSE (a.end - s.location) END * 100\n",
    "        ) / (a.end - a.start)\n",
    ") as percentile\n",
    "from sites_chr1 s, genes_chr1 a WHERE \n",
    "s.strand = a.strand\n",
    "and s.location + (a.end- a.start) >= a.start\n",
    "and s.location - (a.end -a.start) <= a.\"end\"\n",
    "group by percentile order by percentile),\n",
    "\n",
    "percentiles_chr2 as (\n",
    "select Count(*) as sites,\n",
    "        FLOOR(\n",
    "            (\n",
    "                CASE\n",
    "                    WHEN a.strand = '+' then (s.location - a.start)\n",
    "                    ELSE (a.end - s.location) END * 100\n",
    "        ) / (a.end - a.start)\n",
    ") as percentile\n",
    "from sites_chr2 s, genes_chr2 a WHERE \n",
    "s.strand = a.strand\n",
    "and s.location + (a.end- a.start) >= a.start\n",
    "and s.location - (a.end -a.start) <= a.\"end\"\n",
    "group by percentile order by percentile),\n",
    "\n",
    "percentiles_chr3 as(\n",
    "select Count(*) as sites,\n",
    "        FLOOR(\n",
    "            (\n",
    "                CASE\n",
    "                    WHEN a.strand = '+' then (s.location - a.start)\n",
    "                    ELSE (a.end - s.location) END * 100\n",
    "        ) / (a.end - a.start)\n",
    ") as percentile\n",
    "from sites_chr_3 s, genes_chr3 a WHERE \n",
    "s.strand = a.strand\n",
    "and s.location + (a.end- a.start) >= a.start\n",
    "and s.location - (a.end -a.start) <= a.\"end\"\n",
    "group by percentile order by percentile),\n",
    "\n",
    "percentiles_chr4 as (\n",
    "select Count(*) as sites,\n",
    "        FLOOR(\n",
    "            (\n",
    "                CASE\n",
    "                    WHEN a.strand = '+' then (s.location - a.start)\n",
    "                    ELSE (a.end - s.location) END * 100\n",
    "        ) / (a.end - a.start)\n",
    ") as percentile\n",
    "from sites_chr_4 s, genes_chr4 a WHERE \n",
    "s.strand = a.strand\n",
    "and s.location + (a.end- a.start) >= a.start\n",
    "and s.location - (a.end -a.start) <= a.\"end\"\n",
    "group by percentile order by percentile),\n",
    "\n",
    "percentiles_chr5 as (\n",
    "select Count(*) as sites,\n",
    "        FLOOR(\n",
    "            (\n",
    "                CASE\n",
    "                    WHEN a.strand = '+' then (s.location - a.start)\n",
    "                    ELSE (a.end - s.location) END * 100\n",
    "        ) / (a.end - a.start)\n",
    ") as percentile\n",
    "from sites_chr_5 c, genes_chr5 a WHERE \n",
    "s.strand = a.strand\n",
    "and s.location + (a.end- a.start) >= a.start\n",
    "and s.location - (a.end -a.start) <= a.\"end\"\n",
    "group by percentile order by percentile)\n",
    "\n",
    "select percentiles_chr1.sites + percentiles_chr2.sites + percentiles_chr3.sites + percentiles_chr4.sites + percentiles_chr5.sites as sites, percentiles_chr1.percentile as percentile\n",
    "from percentiles_chr1, percentiles_chr2, percentiles_chr3, percentiles_chr4, percentiles_chr5\n",
    "where  \n",
    "percentiles_chr1.percentile = percentiles_chr2.percentile and\n",
    "percentiles_chr1.percentile = percentiles_chr3.percentile and\n",
    "percentiles_chr1.percentile = percentiles_chr4.percentile and\n",
    "percentiles_chr1.percentile = percentiles_chr5.percentile\n",
    "order by percentile\n",
    "    \"\"\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "with genes_chr1 as (Select start, \"end\", strand, chromosome from genes where type = 'gbM' and chromosome = 1),\n",
      "genes_chr2 as (Select start, \"end\", strand, chromosome from genes where type = 'gbM' and chromosome = 2),\n",
      "genes_chr3 as (Select start, \"end\", strand, chromosome from genes where type = 'gbM' and chromosome = 3),\n",
      "genes_chr4 as (Select start, \"end\", strand, chromosome from genes where type = 'gbM' and chromosome = 4),\n",
      "genes_chr5 as (Select start, \"end\", strand, chromosome from genes where type = 'gbM' and chromosome = 5),\n",
      "sites_chr1 as (Select * from sites where context = 'CG' and generation = 0 and chromosome = 1),\n",
      "sites_chr2 as (Select * from sites where context = 'CG' and generation = 0 and chromosome = 2),\n",
      "sites_chr3 as (Select * from sites where context = 'CG' and generation = 0 and chromosome = 3),\n",
      "sites_chr4 as (Select * from sites where context = 'CG' and generation = 0 and chromosome = 4),\n",
      "sites_chr5 as (Select * from sites where context = 'CG' and generation = 0 and chromosome = 5),\n",
      "\n",
      "percentiles_chr1 as (\n",
      "select Count(*) as sites,\n",
      "        FLOOR(\n",
      "            (\n",
      "                CASE\n",
      "                    WHEN a.strand = '+' then (s.location - a.start)\n",
      "                    ELSE (a.end - s.location) END * 100\n",
      "        ) / (a.end - a.start)\n",
      ") as percentile\n",
      "from sites_chr1 s, genes_chr1 a WHERE \n",
      "s.strand = a.strand\n",
      "and s.location + (a.end- a.start) >= a.start\n",
      "and s.location - (a.end -a.start) <= a.\"end\"\n",
      "group by percentile order by percentile),\n",
      "\n",
      "percentiles_chr2 as (\n",
      "select Count(*) as sites,\n",
      "        FLOOR(\n",
      "            (\n",
      "                CASE\n",
      "                    WHEN a.strand = '+' then (s.location - a.start)\n",
      "                    ELSE (a.end - s.location) END * 100\n",
      "        ) / (a.end - a.start)\n",
      ") as percentile\n",
      "from sites_chr2 s, genes_chr2 a WHERE \n",
      "s.strand = a.strand\n",
      "and s.location + (a.end- a.start) >= a.start\n",
      "and s.location - (a.end -a.start) <= a.\"end\"\n",
      "group by percentile order by percentile),\n",
      "\n",
      "percentiles_chr3 as(\n",
      "select Count(*) as sites,\n",
      "        FLOOR(\n",
      "            (\n",
      "                CASE\n",
      "                    WHEN a.strand = '+' then (s.location - a.start)\n",
      "                    ELSE (a.end - s.location) END * 100\n",
      "        ) / (a.end - a.start)\n",
      ") as percentile\n",
      "from sites_chr_3 s, genes_chr3 a WHERE \n",
      "s.strand = a.strand\n",
      "and s.location + (a.end- a.start) >= a.start\n",
      "and s.location - (a.end -a.start) <= a.\"end\"\n",
      "group by percentile order by percentile),\n",
      "\n",
      "percentiles_chr4 as (\n",
      "select Count(*) as sites,\n",
      "        FLOOR(\n",
      "            (\n",
      "                CASE\n",
      "                    WHEN a.strand = '+' then (s.location - a.start)\n",
      "                    ELSE (a.end - s.location) END * 100\n",
      "        ) / (a.end - a.start)\n",
      ") as percentile\n",
      "from sites_chr_4 s, genes_chr4 a WHERE \n",
      "s.strand = a.strand\n",
      "and s.location + (a.end- a.start) >= a.start\n",
      "and s.location - (a.end -a.start) <= a.\"end\"\n",
      "group by percentile order by percentile),\n",
      "\n",
      "percentiles_chr5 as (\n",
      "select Count(*) as sites,\n",
      "        FLOOR(\n",
      "            (\n",
      "                CASE\n",
      "                    WHEN a.strand = '+' then (s.location - a.start)\n",
      "                    ELSE (a.end - s.location) END * 100\n",
      "        ) / (a.end - a.start)\n",
      ") as percentile\n",
      "from sites_chr_5 c, genes_chr5 a WHERE \n",
      "s.strand = a.strand\n",
      "and s.location + (a.end- a.start) >= a.start\n",
      "and s.location - (a.end -a.start) <= a.\"end\"\n",
      "group by percentile order by percentile)\n",
      "\n",
      "select percentiles_chr1.sites + percentiles_chr2.sites + percentiles_chr3.sites + percentiles_chr4.sites + percentiles_chr5.sites as sites, percentiles_chr1.percentile as percentile\n",
      "from percentiles_chr1, percentiles_chr2, percentiles_chr3, percentiles_chr4, percentiles_chr5\n",
      "where  \n",
      "percentiles_chr1.percentile = percentiles_chr2.percentile and\n",
      "percentiles_chr1.percentile = percentiles_chr3.percentile and\n",
      "percentiles_chr1.percentile = percentiles_chr4.percentile and\n",
      "percentiles_chr1.percentile = percentiles_chr5.percentile\n",
      "order by percentile\n",
      "    \n"
     ]
    }
   ],
   "source": [
    "print(meth_sites_query('gbM', 'CG'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "import clickhouse_connect\n",
    "import polars as pl\n",
    "client = clickhouse_connect.get_client(host='localhost', username='cgoeldel', password='Goe1409ldel')\n",
    "\n",
    "chr_results = {}\n",
    "\n",
    "for gene_type in [\"UM\", \"teM\", \"gbM\"]:\n",
    "    for state in range(1, 36 + 1):\n",
    "        res = client.query_arrow(chr_state_query(gene_type, state))\n",
    "        df = pl.DataFrame(res)\n",
    "        df.write_csv(f\"/mnt/fast/epigenomics/chr_states/percentiles/{gene_type}_{state}.csv\")\n",
    "        chr_results[f\"{gene_type}_{state}\"] = df\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot the results\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "\n",
    "gbM_results = np.zeros((36, 300))\n",
    "UM_results = np.zeros((36, 300))\n",
    "teM_results = np.zeros((36, 300))\n",
    "\n",
    "\n",
    "\n",
    "for (i, df) in chr_results.items():\n",
    "    gene_type, state = i.split(\"_\")\n",
    "    for row in df.iter_rows(named=True):\n",
    "        idx = int(row[\"percentile\"]) +100\n",
    "        if gene_type == \"gbM\":\n",
    "            gbM_results[int(state) - 1, idx] = row[\"sites\"]\n",
    "        elif gene_type == \"UM\":\n",
    "            UM_results[int(state) - 1, idx] = row[\"sites\"]\n",
    "        elif gene_type == \"teM\":\n",
    "            teM_results[int(state) - 1, idx] = row[\"sites\"]\n",
    "\n",
    "gbM_results = pl.from_numpy(gbM_results, schema=[f\"S{i+1}\" for i in range(36)])\n",
    "UM_results = pl.from_numpy(UM_results, schema=[f\"S{i+1}\" for i in range(36)])\n",
    "teM_results = pl.from_numpy(teM_results, schema=[f\"S{i+1}\" for i in range(36)])\n",
    "\n",
    "gbM_results.write_csv(\"/mnt/fast/epigenomics/chr_states/percentiles/gbM.csv\")\n",
    "UM_results.write_csv(\"/mnt/fast/epigenomics/chr_states/percentiles/UM.csv\")\n",
    "teM_results.write_csv(\"/mnt/fast/epigenomics/chr_states/percentiles/teM.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "def q(type=\"gbM\"):\n",
    "    return f\"\"\"\n",
    "with percentile_windows as (\n",
    "    select Count(*) as sites,\n",
    "        FLOOR(\n",
    "            (\n",
    "                CASE\n",
    "                    WHEN a.strand = '+' then ((c.start +c.end)/2 - a.start)\n",
    "                    else (a.end - (c.start + c.end)/2\n",
    "                )\n",
    "            end * 100\n",
    "        ) / (a.end - a.start\n",
    ")\n",
    ") as percentile, c.state as state\n",
    "from chr_states c\n",
    "    JOIN genes a ON c.chromosome = a.chromosome\n",
    "where (c.start +c.end)/2 + (a.end - a.start\n",
    ") >= a.start\n",
    "and (c.start +c.end)/2 - (a.end - a.start\n",
    ") <= a.\"end\" and a.type = '{type}'\n",
    "group by percentile,  state order by state, percentile\n",
    ")\n",
    "select sites, state,\n",
    "    SUM(sites) OVER (\n",
    "        ORDER BY percentile ROWS BETWEEN CURRENT ROW\n",
    "            AND 4 FOLLOWING\n",
    "    ) AS sites_sliding_sum,\n",
    "    percentile\n",
    "from percentile_windows order by state, percentile\n",
    "\"\"\"\n",
    "\n",
    "res = client.query_arrow(q(\"gbM\"))\n",
    "df = pl.DataFrame(res)\n",
    "\n",
    "res = client.query_arrow(q(\"UM\"))\n",
    "df2 = pl.DataFrame(res)\n",
    "\n",
    "res = client.query_arrow(q(\"teM\"))\n",
    "df3 = pl.DataFrame(res)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [],
   "source": [
    "gbM = df.pivot(columns=\"state\", values=\"sites\", index=\"percentile\" ).sort(\"percentile\").fill_null(0)\n",
    "UM = df2.pivot(columns=\"state\", values=\"sites\", index=\"percentile\" ).sort(\"percentile\").fill_null(0)\n",
    "teM = df3.pivot(columns=\"state\", values=\"sites\", index=\"percentile\" ).sort(\"percentile\").fill_null(0)\n",
    "\n",
    "gbM.write_csv(\"/mnt/fast/epigenomics/chr_states/percentiles/gbM_pivot.csv\")\n",
    "UM.write_csv(\"/mnt/fast/epigenomics/chr_states/percentiles/UM_pivot.csv\")\n",
    "teM.write_csv(\"/mnt/fast/epigenomics/chr_states/percentiles/teM_pivot.csv\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".env",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
